Content.Shared/Research/Systems/SharedResearchSystem.cs

```diff
@@ -1,46 +1,147 @@
-﻿using Content.Shared.Research.Components;
+﻿using System.Linq;
+using Content.Shared.Research.Components;
 using Content.Shared.Research.Prototypes;
-using Robust.Shared.GameStates;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
 
 namespace Content.Shared.Research.Systems;
 
 public abstract class SharedResearchSystem : EntitySystem
 {
+    [Dependency] protected readonly IPrototypeManager PrototypeManager = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+
     public override void Initialize()
     {
         base.Initialize();
 
-        SubscribeLocalEvent<ResearchServerComponent, ComponentGetState>(OnServerGetState);
-        SubscribeLocalEvent<ResearchServerComponent, ComponentHandleState>(OnServerHandleState);
-        SubscribeLocalEvent<TechnologyDatabaseComponent, ComponentGetState>(OnTechnologyGetState);
-        SubscribeLocalEvent<TechnologyDatabaseComponent, ComponentHandleState>(OnTechnologyHandleState);
+        SubscribeLocalEvent<TechnologyDatabaseComponent, MapInitEvent>(OnMapInit);
     }
 
-    private void OnServerGetState(EntityUid uid, ResearchServerComponent component, ref ComponentGetState args)
+    private void OnMapInit(EntityUid uid, TechnologyDatabaseComponent component, MapInitEvent args)
     {
-        args.State = new ResearchServerState(component.ServerName, component.Points, component.Id);
+        UpdateTechnologyCards(uid, component);
     }
 
-    private void OnServerHandleState(EntityUid uid, ResearchServerComponent component, ref ComponentHandleState args)
+    public void UpdateTechnologyCards(EntityUid uid, TechnologyDatabaseComponent? component = null)
     {
-        if (args.Current is not ResearchServerState state)
+        if (!Resolve(uid, ref component))
             return;
-        component.ServerName = state.ServerName;
-        component.Points = state.Points;
-        component.Id = state.Id;
+
+        var availableTechnology = GetAvailableTechnologies(uid, component);
+
+        component.CurrentTechnologyCards.Clear();
+        foreach (var discipline in component.SupportedDisciplines)
+        {
+            var filtered = availableTechnology.
+                Where(p => p.Discipline == discipline &&
+                           !component.UnlockedTechnologies.Contains(p.ID)).ToList();
+
+            if (!filtered.Any())
+                continue;
+
+            component.CurrentTechnologyCards.Add(_random.Pick(filtered).ID);
+        }
```

Comment:
imo makes more sense to just shuffle `availableTechnology` then iterate over it and pick the first of each discipline thats found. `!component.UnlockedTechnologies.Contains(p.ID)` shouldnt be necessary since `IsTechnologyAvailable` already checks that